diff --git a/src/lib/Hydra/Helper/Nix.pm b/src/lib/Hydra/Helper/Nix.pm
index 71a8a7d7..b6a9eb19 100644
--- a/src/lib/Hydra/Helper/Nix.pm
+++ b/src/lib/Hydra/Helper/Nix.pm
@@ -180,6 +180,8 @@ sub getDrvLogPath {
 sub findLog {
     my ($c, $drvPath, @outPaths) = @_;

+    @outPaths = grep {defined} @outPaths;
+
     if (defined $drvPath) {
         my $logPath = getDrvLogPath($drvPath);
         return $logPath if defined $logPath;
