diff --git a/src/libstore/daemon.cc b/src/libstore/daemon.cc
index de69b50ee..f64874b58 100644
--- a/src/libstore/daemon.cc
+++ b/src/libstore/daemon.cc
@@ -174,7 +174,7 @@ struct TunnelSource : BufferedSource
         to << STDERR_READ << len;
         to.flush();
         size_t n = readString(data, len, from);
-        if (n == 0) throw EndOfFile("unexpected end-of-file");
+        if (n == 0) throw EndOfFile("TunnelSource::readUnbuffered: unexpected end-of-file");
         return n;
     }
 };
diff --git a/src/libutil/serialise.cc b/src/libutil/serialise.cc
index 8ff904583..7348ad35a 100644
--- a/src/libutil/serialise.cc
+++ b/src/libutil/serialise.cc
@@ -143,7 +143,7 @@ size_t FdSource::readUnbuffered(char * data, size_t len)
         n = ::read(fd, data, len);
     } while (n == -1 && errno == EINTR);
     if (n == -1) { _good = false; throw SysError("reading from file"); }
-    if (n == 0) { _good = false; throw EndOfFile("unexpected end-of-file"); }
+    if (n == 0) { _good = false; throw EndOfFile("FdSource::readUnbuffered: unexpected end-of-file"); }
     read += n;
     return n;
 }
diff --git a/src/libutil/util.cc b/src/libutil/util.cc
index 1c19938a8..87aef9a54 100644
--- a/src/libutil/util.cc
+++ b/src/libutil/util.cc
@@ -689,7 +689,7 @@ void readFull(int fd, char * buf, size_t count)
             if (errno == EINTR) continue;
             throw SysError("reading from file");
         }
-        if (res == 0) throw EndOfFile("unexpected end-of-file");
+        if (res == 0) throw EndOfFile("readFull: unexpected end-of-file");
         count -= res;
         buf += res;
     }
